### DISCUSSION TITLE
MaxListenersExceededWarning: Possible EventEmitter memory leak detected.

### DISCUSSION THREAD

**Discussion by janober** [pyodide/pyodide#3541] [General]
Hopefully my last problem.

For our use case, will people run pyodide potentially very often. So I did run a test and running `await loadPyodide()` multiple times results in the following error after a few iterations:
```
(node:38) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
(node:38) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 unhandledRejection listeners added to [process]. Use emitter.setMaxListeners() to increase limit
```
Is there some close/cleanup/destroy call required that I am missing. Was sadly not able to find anything in the docs. Once that issue started to appear I thought it could be related to the PyProxy as I saw a comment about that. After debugging that for a while, did I realize that it also happens even without the `pyodide.runPythonAsync` call and just the above-mentioned `loadPyodide()`.

Am I missing something? Thanks a lot!

**Comment by hoodmane** [comment]
Loading multiple copies of Pyodide is still a somewhat experimental feature (first release supporting it was 0.21.0 which was six months ago). As you have correctly noticed, there is no proper lifecycle support. Currently there is the fatal error handling but no support for interpreter shutdown (so for instance `atexit` handlers are never called). We should add a `shutdown()` api that calls `Py_FinalizeEx`, detaches the backing memory, and as best as possible ensures that everything can be garbage collected.

This warning is caused by an upstream issue: emscripten doesn't have true lifecycle support either. The handler is added here
https://github.com/emscripten-core/emscripten/blob/main/src/shell.js#L234
and there is no way to remove it. We could compile with `-sNODEJS_CATCH_EXIT=0` and that code would not be added. If we want we could add our own version of that logic that can actually be removed at exit.

We could add a test that repeatedly calls `loadPyodide` and then shuts it down and checks that there is no memory error or other noticeable leaks. Another possibility is to add a `FinalizationRegistry` to automatically do this cleanup if a `pyodide` is gc'd.

One other basic issue is that the instance checks for `PyProxy` are not correctly tied to a single interpreter so if you load two interpreters and move a `PyProxy` from one to the other you will get memory errors.

**Comment by hoodmane** [comment]
You can suppress the warning by calling:
```js
process.removeAllListeners("uncaughtException");
```
after each `loadPyodide` call.

Depending on your use case using workers might work better:
https://nodejs.org/api/worker_threads.html#worker-threads
They can be a bit of a pain due to the message passing API though. Of course another thing which may well not be an option is to restart node.

**Reply by hoodmane** [reply]
Actually it looks like:
```js
process.removeListener("uncaughtException", process.rawListeners("uncaughtException").pop());
```
should do the trick without changing global state as badly. I think this listener provides no useful functionality since Pyodide will always catch the `ExitStatus` object anyways.

**Reply by hoodmane** [reply]
Seems that you also have to filter `unhandledRejection`.

**Comment by janober** [comment]
As always, thanks a lot for the amazing fast help!

Actually realized I do not really have to import pyodide multiple times. Rather I could make it a singleton, and now it seems to work fine. At least assuming it is fine throwing a lot of data at one instance does not cause other issues.

**Reply by hoodmane** [reply]
I believe most people use a single Pyodide instance for a lot of different tasks and generally don't report too many problems. One issue people sometimes have is if they modify a Python module they have trouble getting a new one. Also if you need to run `atexit` handlers and other stuff that happens at shutdown, that would be a problem. You can force this with `pyodide._module._Py_FinalizeEx()` and then you need to make a new Pyodide instance.

You can use `runPyodide(code, {globals: someDictProxy});` which will isolate the global name spaces of different tasks if you don't want them to step on each other. Or write code to a file and use `pyodide.pyimport`.

There are also private `saveState` and `restoreState` APIs that restore most things.   
```js
let state = pyodide._api.saveState();
// ... do stuff
pyodide._api.restoreState(state);
```
These are primarily intended for our own testing purposes but I have heard that some people are using them.

**Comment by janober** [comment]
Thanks, you actually read my mind! That is exactly something that came then also to my mind. `runPyodide` seems to expect a `PyProxy`. Very stupid question, but how can I create one in JS?

**Reply by hoodmane** [reply]
```js
const dict = pyodide.globals.get("dict");
let tempDict = dict();
pyodide.runPython(code, {globals: tempDict});
// When you're done with it
tempDict.destroy();
```

**Reply by janober** [reply]
Amazing, thanks a lot. Works great.
